### PR TITLE
Cache mini cart frontend dependencies

### LINKED ISSUE #52000: [Enhancement]: Mini cart lazy loaded scripts are performance heavy
### Describe the solution you'd like

Getting all the lazy loaded scripts and translations for the mini cart is somewhat time consuming.  We should consider caching the result of these values and adding the cached value inline when available.

https://github.com/woocommerce/woocommerce/blob/update/51953/plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php/#L92

### Describe alternatives you've considered

It's possible to statically type out all the dependencies, but this isn't very maintenance friendly.  A cached solution is much easier and nearly identical in terms of performance.

### Additional context

Related to https://github.com/woocommerce/woocommerce/issues/50714.

### PR DESCRIPTION
**Pull request by joshuatf**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Caches the mini cart dependencies for empty and non-empty carts to improve performance.  This saves up to 2ms in my local environment (a 6.8% improvement over baseline in #50714).

Screenshot 2024-10-11 at 11 57 46 AM

Closes #52000 .

#### Regression Testing

1. Make sure you have a block theme activated and the mini cart is present in the header
3. Navigate to the homepage of your site
4. Open your console, type `wcBlocksMiniCartFrontendDependencies` and hit enter
5. Make sure the following dependencies screenshot matches when no items are in the cart and when items are in the cart
6. In the frontend, open the Network tab of your browser devtools and verify the Mini Cart inner blocks are being preloaded (you can search for empty-cart-frontend or filled-cart-frontend).
7. Interact with the Mini Cart and verify there are no regressions.

(Empty cart)
Screenshot 2024-10-11 at 12 00 41 PM
(Cart with items)
Screenshot 2024-10-11 at 11 58 50 AM

#### Performance Testing

1. Make sure you have `WP_DEBUG_LOG` enabled in your wp config.
2. Check out the branch `update/52000-performance`
3. In your `wp_options` table, delete the `_site_transient_woocommerce_mini_cart_empty_frontend_dependencies` and `_site_transient_woocommerce_mini_cart_non_empty_frontend_dependencies` options
4. Navigate to your homepage and reload a couple times
5. Open your debug log and compare the performance times with and without cache

#### Development Testing

1. Navigate to the homepage of your site with the mini cart block enabled
2. Remove a dependency from the mini cart dependencies array (e.g. https://github.com/woocommerce/woocommerce/blob/update/52000-performance/plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php/#L272)
3. Reload the page and open your console and verify that the value of `wcBlocksMiniCartFrontendDependencies` still has the dependency
4. Bump the version of WC on this line - https://github.com/woocommerce/woocommerce/blob/update/52000-performance/plugins/woocommerce/includes/class-woocommerce.php/#L48
5. Reload the page and open your console and verify that the value of `wcBlocksMiniCartFrontendDependencies` is now updated to reflect your dependency list
6. In your `wp-config.php` file, add `define( 'WP_DEVELOPMENT_MODE', 'plugin' );`
7. Once again, modify the dependency list in https://github.com/woocommerce/woocommerce/blob/update/52000-performance/plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php/#L272
8. Check that changes to `wcBlocksMiniCartFrontendDependencies` are immediately reflected without needing to bump the version number of WC or WP

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/changelog/update-52000
Comment (Aljullu) [inline plugins/woocommerce/changelog/update-52000]: Nit: given that this is added to the release post, I would use the block user-friendly name:

```suggestion
Cache Mini-Cart block frontend dependencies
```

**Review by Aljullu — COMMENTED** [review]
Thanks for working on this, @joshuatf! It's definitely a great performance improvement! :clap: 

Code changes look good to me, but I see some tests are failing related to translations and a nonce issue. Let me know if you would like me to look into this (it would need to be next week, though, as I'm in support rotation).

I think we should also ask for a review from Rubik, as they are the ones maintaining the Mini-Cart block.

File: plugins/woocommerce/changelog/update-52000
Comment (joshuatf) [reply plugins/woocommerce/changelog/update-52000]: Good call!  Updated.

**Comment by joshuatf** [comment]
Thanks for taking a look, @Aljullu!

>Code changes look good to me, but I see some tests are failing related to translations and a nonce issue. 

Managed to get these fixed.  The workaround is a tiny bit hacky, but we reappend before/after scripts so that the nonce does not get cached in https://github.com/woocommerce/woocommerce/pull/52001/commits/aa92e3fd33f64c2298b66af1990b31f7498efa9b.

Also had to run the caching on shutdown as there seems to be a race condition with the script data caching that also occurs on shutdown.  That should be fixed as of https://github.com/woocommerce/woocommerce/pull/52001/commits/5d74f3b13e020e513ffeae3294fbc49e09cf46fb.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: Nit: I know this wasn't introduced in this PR, but [we don't support WP 6.3 or lower](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/woocommerce.php#L11) anymore, so it might be a good opportunity to clean this up.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: I think this function isn't used anywhere? :thinking:

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: Do you think we should add a check to verify `$cache['data']` exists and it's a string? Maybe I'm being unnecessarily cautious, but given that it's a value that comes from the database, users or other plugins might have messed up with its format.

**Review by Aljullu — APPROVED** [review]
Thanks for the updates, @joshuatf! I did some more testing and everything seems to be working well on my end. Code changes also look good. I left a few comments, but I'm pre-approving as all of them are minor. However, feel free to ping me if you would like another review.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (joshuatf) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: Good call!  Removed.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (joshuatf) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: 🤦  Accidentally left over from first iteration.  Removed.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (joshuatf) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php]: I think that's a reasonable precaution to take.  Added some checks in https://github.com/woocommerce/woocommerce/pull/52001/commits/19276cb0ec99be82f98327b3e9fdef7af47b724a.

**Comment by joshuatf** [comment]
Thanks for the review and great suggestions, @Aljullu!  Made all of those changes.

It would be good to get another set of eyes from someone on @woocommerce/rubik (cc @senadir) since they maintain this.

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (alexflorisca) [inline plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: I use `define( 'WP_ENVIRONMENT_TYPE', 'local' );` which I can't remember what for now but may be something to consider adding here

**Review by alexflorisca — COMMENTED** [review]
Thanks for the ping, just took a look at this.

There's a bunch of dependencies in the `wcBlocksMiniCartFrontendDependencies` array that aren't here for me, but I don't think they are related to this PR? They are things like`wp-warning`, `wp-style-engine` and a few others. The two that might be relevant are `wc-blocks-checkout` and `wc-blocks-components`?

> Remove a dependency from the mini cart dependencies array (e.g. https://github.com/woocommerce/woocommerce/blob/update/52000-performance/plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php/#L272)
Reload the page and open your console and verify that the value of wcBlocksMiniCartFrontendDependencies still has the dependency

Interestingly, this doesn't work for empty cart but does work for the filled cart. If I remove anything from the first array (empty cart), they also disappear from the frontend global. 

The rest seems ok to me. I did some loose performance testing (just measuring  and with caching page load was about 1.66, without it was 1.84. So pretty significant change there!

File: plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php
Comment (senadir) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: I don't think we should be the one defining `WP_ENVIRONMENT_TYPE`, it comes from WP-config

Comment (alexflorisca) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: not defining, just checking it

Comment (senadir) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: Ah yeah that's what `wp_is_development_mode` is doing here + other checks and filters

https://developer.wordpress.org/reference/functions/wp_get_development_mode/

Comment (alexflorisca) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: Doesn't look like it checks `WP_ENVIRONMENT_TYPE` only `WP_DEVELOPMENT_MODE`. They seem like they do similar things so using that function is probably good enough.

Comment (senadir) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: ah good point, I didn’t know there exists 2, sorry.

Comment (joshuatf) [reply plugins/woocommerce/src/Blocks/BlockTypes/MiniCart.php:326]: This is relatively new ([available since WP 6.3](https://make.wordpress.org/core/2023/07/14/configuring-development-mode-in-6-3/)) and differs slightly from the previous `WP_ENVIRONMENT_TYPE`, `WP_DEBUG`, and `SCRIPT_DEBUG`.

The others relate to other aspects of the environment, while `wp_is_development_mode` is commonly used for asset caching in WP core.

**Comment by joshuatf** [comment]
Thanks for the review, @alexflorisca!

>Interestingly, this doesn't work for empty cart but does work for the filled cart. If I remove anything from the first array (empty cart), they also disappear from the frontend global.

I think subsequent requests may be overwriting the cache and forcing it to update each page.  I updated this behavior to only update when lazy loaded scripts are instantiated in https://github.com/woocommerce/woocommerce/pull/52001/commits/18a8c58c4d3e81f4011ad46c34d7293cf79e5104.

Could you let me know if that works for you, @alexflorisca?

**Review by alexflorisca — APPROVED** [review]
Works great now, thanks 👍

**Comment by alexflorisca** [comment]
Just noting that the Mini Cart translation test keeps failing, probably related to this PR. Would be great if you can take a look before merging

**Comment by joshuatf** [comment]
>Just noting that the Mini Cart translation test keeps failing, probably related to this PR. Would be great if you can take a look before merging

It is related to this PR and started failing in https://github.com/woocommerce/woocommerce/pull/52001/commits/194aaa2427d5c0ed1f01712c55ff3f3a6f81f3db.

I'll make sure this is resolved before merging and ping for re-review if any changes are introduced.

**Comment by joshuatf** [comment]
After spending some time on this, it seems that caching can introduce complications with various script dependencies.  While I had originally worked around this by manually injecting dependencies into the cache on each load, the list has grown to the point that it eats into the performance gains in this PR.

Given the now relatively small performance gains on this PR and the overhead it adds, I'm going to close this PR.

Thanks to everyone who took the time to review 😄

**Pull request closed without merging**
